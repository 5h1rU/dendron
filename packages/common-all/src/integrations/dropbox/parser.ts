import { ListFolderResultSimple, MetadataReference } from "./types";
import { dxId2DendronId, fileNameToTreePath } from "./utils";

import { IDNode } from "../../types";
import { Note } from "../../node";
import _ from "lodash";

interface PathEnt<T> {
  path: string[];
  data: T;
  size: number
}

type PathDict<T> = { [key: string]: PathEnt<T> };

class DropboxTreeParser {
  public seed: ListFolderResultSimple;
  public pathDict: PathDict<MetadataReference>;

  constructor(seed: ListFolderResultSimple) {
    this.seed = seed;
    this.pathDict = {};
    this.seedToPathDict();
  }

  seedToPathDict() {
    this.seed.entries.forEach(ent => {
      const { name } = ent;
      const treePath = fileNameToTreePath(name);
      const size = treePath.length
      this.pathDict[treePath.join(".")] = { path: treePath, data: ent, size };
    });
  }

  execute(): Note[] {
    const currentLvl = _.pickBy(this.pathDict, {size: 1})

    prefixes where {size: 1}
    
    const nodes = this.getTopLevelNodes();
    nodes.forEach(n => {
      this.parse(n);
    });
    return nodes;
  }

  parse(node: Note) {}
}
