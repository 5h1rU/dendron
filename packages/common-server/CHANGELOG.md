# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.3.24](https://github.com/dendronhq/dendron/compare/v0.3.23...v0.3.24) (2020-07-20)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.21](https://github.com/dendronhq/dendron/compare/v0.3.20...v0.3.21) (2020-07-19)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.20](https://github.com/dendronhq/dendron/compare/v0.3.19...v0.3.20) (2020-07-18)

### Bug Fixes

- handle special characters in file names when importing using LocalFilePod ([03e42b1](https://github.com/dendronhq/dendron/commit/03e42b167ac9f073cd56f10c4e31b5cecf66dabf))

## [0.3.19](https://github.com/dendronhq/dendron/compare/v0.3.18...v0.3.19) (2020-07-18)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.18](https://github.com/dendronhq/dendron/compare/v0.3.17...v0.3.18) (2020-07-18)

### Features

- ability to import notes from local file system [docs](https://github.com/dendronhq/dendron-template/blob/master/vault/dendron.topic.pod.md) ([e44fd11](https://github.com/dendronhq/dendron/commit/e44fd1133bec10f22831f059c8d98cf4076dcdcc))

## [0.3.17](https://github.com/dendronhq/dendron/compare/v0.3.16...v0.3.17) (2020-07-17)

### Features

- more flexible file names for [journal notes](https://github.com/dendronhq/dendron-template/blob/master/vault/dendron.special-notes.md) ([9d9f10b](https://github.com/dendronhq/dendron/commit/9d9f10bd0873c201361c4625fc49c9f62ee82991))

## [0.3.16](https://github.com/dendronhq/dendron/compare/v0.3.15...v0.3.16) (2020-07-16)

### Features

- implement journal notes. see details here: https://github.com/dendronhq/dendron-template/blob/master/vault/dendron.feature.journals.md ([5e1236f](https://github.com/dendronhq/dendron/commit/5e1236fddbf1e0fddf4c27d1a40e9841cc99974f))

## [0.3.15](https://github.com/dendronhq/dendron/compare/v0.3.14...v0.3.15) (2020-07-14)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.11](https://github.com/dendronhq/dendron/compare/v0.3.10...v0.3.11) (2020-07-13)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.10](https://github.com/dendronhq/dendron/compare/v0.1.6...v0.3.10) (2020-07-13)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.9](https://github.com/dendronhq/dendron/compare/v0.3.8...v0.3.9) (2020-07-09)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.8](https://github.com/dendronhq/dendron/compare/v0.3.7...v0.3.8) (2020-07-09)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.7](https://github.com/dendronhq/dendron/compare/v0.3.6...v0.3.7) (2020-07-08)

**Note:** Version bump only for package @dendronhq/common-server

## [0.3.6](https://github.com/dendronhq/dendron/compare/v0.3.5...v0.3.6) (2020-07-07)

**Note:** Version bump only for package @dendronhq/common-server

# [0.3.0](https://github.com/dendronhq/dendron/compare/v0.2.20...v0.3.0) (2020-07-05)

**Note:** Version bump only for package @dendronhq/common-server
