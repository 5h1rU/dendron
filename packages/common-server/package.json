{
  "name": "@dendronhq/common-server",
  "version": "0.1.6",
  "description": "common-server",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "ssh://git@github.com/dendronhq/dendron.git",
    "directory": "packages/common-server"
  },
  "author": {
    "name": "Kevin Lin",
    "email": "kevin@thence.io",
    "url": "https://kevinslin.com",
    "organization": true
  },
  "keywords": [
    ""
  ],
  "files": [
    "lib"
  ],
  "main": "lib/index",
  "types": "lib/index",
  "scripts": {
    "clean": "rimraf lib && rimraf coverage",
    "format": "",
    "lint": "",
    "prepublishOnly": "npm run build",
    "prebuild": "npm run clean && npm run format && npm run lint && echo Using TypeScript && tsc --version",
    "build": "npm run compile",
    "compile": "tsc -p tsconfig.build.json",
    "test": "jest",
    "coverage": "jest --coverage",
    "watch": "npm run compile -- --watch",
    "test:watch": "jest --watch"
  },
  "dependencies": {
    "@dendronhq/common-all": "^0.1.6",
    "gray-matter": "^4.0.2",
    "lodash": "^4.17.15",
    "minimatch": "^3.0.4",
    "pino": "^6.3.2",
    "yamljs": "^0.3.0"
  },
  "devDependencies": {
    "@types/pino": "^6.3.0",
    "@types/jest": "^23.3.3",
    "@types/lodash": "^4.14.154",
    "@types/minimatch": "^3.0.3",
    "@types/node": "^10.11.4",
    "@types/yamljs": "^0.2.31",
    "coveralls": "^3.0.2",
    "jest": "^23.6.0",
    "prettier": "^1.14.3",
    "rimraf": "^2.6.2",
    "ts-jest": "^23.10.3",
    "ts-node": "^7.0.1",
    "tslint": "^5.11.0",
    "tslint-config-prettier": "^1.15.0",
    "typescript": "^3.1.1"
  },
  "engines": {
    "node": ">=10.0.0"
  },
  "jest": {
    "preset": "ts-jest"
  },
  "gitHead": "403eb012f8b34bc115dd7dd86f80c76b39329f3b"
}
