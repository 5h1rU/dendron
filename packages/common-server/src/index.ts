export * from "./files";
export * from "./logger";
export { testUtils, FileTestUtils } from "./testUtils";
