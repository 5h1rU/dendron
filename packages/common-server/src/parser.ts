import {
  DendronError,
  DNodeUtils,
  DStoreV2,
  DVault,
  ENGINE_ERROR_CODES,
  SchemaModuleOpts,
  SchemaModuleProps,
  SchemaOpts,
  SchemaPropsDict,
  SchemaProps,
  SchemaUtils,
} from "@dendronhq/common-all";
import _ from "lodash";
import path from "path";
import { file2Schema, vault2Path } from "./filesv2";
import { DLogger } from "./logger";
import { createLogger } from "./logger";

let _LOGGER: DLogger | undefined;

function getLogger() {
  if (!_LOGGER) {
    _LOGGER = createLogger();
  }
  return _LOGGER;
}

export class ParserBaseV2 {
  constructor(public opts: { store: DStoreV2; logger: any }) {}

  get logger() {
    return this.opts.logger;
  }
}

export class SchemaParserV2 extends ParserBaseV2 {
  static parseRaw(
    schemaOpts: SchemaModuleOpts,
    opts: { root: DVault; fname: string; wsRoot: string }
  ): SchemaModuleProps {
    const version = _.isArray(schemaOpts) ? 0 : 1;
    if (version > 0) {
      return SchemaParserV2.parseSchemaModuleOpts(
        schemaOpts as SchemaModuleOpts,
        opts
      );
    } else {
      // TODO: legacy
      const schemaDict: SchemaPropsDict = {};
      ((schemaOpts as unknown) as SchemaOpts[]).map((ent) => {
        const schema = SchemaUtils.create(ent);
        schemaDict[schema.id] = schema;
      });
      const maybeRoot = _.find(_.values(schemaDict), {
        parent: "root",
      }) as SchemaProps;
      return {
        version: 0,
        root: maybeRoot,
        schemas: schemaDict,
        fname: opts.fname,
        vault: opts.root,
      };
    }
  }

  static parseSchemaModuleOpts(
    schemaModuleProps: SchemaModuleOpts,
    opts: { fname: string; root: DVault; wsRoot: string }
  ): SchemaModuleProps {
    const { imports, schemas, version } = schemaModuleProps;
    const { fname, root, wsRoot } = opts;
    getLogger().info({ ctx: "parseSchemaModuleOpts", fname, root, imports });
    const vpath = vault2Path({ vault: root, wsRoot });
    let schemaModulesFromImport = _.flatMap(imports, (ent) => {
      const fpath = path.join(vpath, ent + ".schema.yml");
      return file2Schema(fpath, wsRoot);
    });
    const schemaPropsFromImport = schemaModulesFromImport.flatMap((mod) => {
      const domain = mod.fname;
      return _.values(mod.schemas).map((ent) => {
        ent.data.pattern = ent.data.pattern || ent.id;
        ent.id = `${domain}.${ent.id}`;
        ent.fname = fname;
        ent.parent = null;
        ent.children = ent.children.map((ent) => `${domain}.${ent}`);
        ent.vault = root;
        return ent;
      });
    });
    getLogger().debug({ ctx: "parseSchemaModuleOpts", schemaPropsFromImport });
    const schemaPropsFromFile = schemas.map((ent) => {
      return SchemaUtils.create({ ...ent, vault: root });
    });
    getLogger().debug({ ctx: "parseSchemaModuleOpts", schemaPropsFromFile });
    const schemasAll = schemaPropsFromImport.concat(schemaPropsFromFile);

    const schemasDict: SchemaPropsDict = {};
    schemasAll.forEach((ent) => {
      schemasDict[ent.id] = ent;
    });

    const rootModule = SchemaUtils.getModuleRoot(schemaModuleProps);

    const addConnections = (parent: SchemaProps) => {
      _.map(parent.children, (ch) => {
        const child = schemasDict[ch];
        if (!child) {
          throw new DendronError({
            status: ENGINE_ERROR_CODES.MISSING_SCHEMA,
            msg: JSON.stringify({ parent, missingChild: ch }),
          });
        }
        DNodeUtils.addChild(parent, child);
        return addConnections(child);
      });
    };
    // add parent relationship
    addConnections(rootModule);

    return {
      version,
      imports,
      root: rootModule,
      schemas: schemasDict,
      fname,
      vault: root,
    };
  }
}
