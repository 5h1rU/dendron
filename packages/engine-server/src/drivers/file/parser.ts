import {
  DEngineStore,
  DNodeUtils,
  Note,
  NoteDict,
  NoteRawProps,
  NoteUtils,
  SchemaNodeRaw,
  SchemaRawOpts,
  SchemaRawProps
} from "@dendronhq/common-all";
import {
  createLogger,
  globMatch,
  mdFile2NodeProps
} from "@dendronhq/common-server";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";
import YAML from "yamljs";

// @ts-ignore
const logger = createLogger("FileParser");

export type FileMeta = {
  // file name: eg. foo.md, name = foo
  prefix: string;
  // fpath: full path, eg: foo.md, fpath: foo.md
  fpath: string;
};
type FileMetaDict = { [key: string]: FileMeta[] };

function getFileMeta(fpaths: string[]): FileMetaDict {
  const metaDict: FileMetaDict = {};
  _.forEach(fpaths, fpath => {
    const { name } = path.parse(fpath);
    const lvl = name.split(".").length;
    if (!_.has(metaDict, lvl)) {
      metaDict[lvl] = [];
    }
    metaDict[lvl].push({ prefix: name, fpath });
  });
  return metaDict;
}

type FileParserOpts = {
  errorOnEmpty?: boolean;
  errorOnBadParse?: boolean;
};
type FileParserProps = Required<FileParserOpts>;

export class FileParser {
  public errors: any[];

  public opts: FileParserProps;

  public missing: Set<string>;

  public store: DEngineStore;

  constructor(store: DEngineStore, opts?: FileParserOpts) {
    this.errors = [];
    this.missing = new Set<string>();
    this.opts = _.defaults(opts, {
      errorOnEmpty: true,
      errorOnBadParse: true
    });
    this.store = store;
  }

  toNode(
    ent: FileMeta,
    parents: Note[],
    store: DEngineStore,
    opts?: {
      errorOnEmpty?: boolean;
      isRoot?: boolean;
      errorOnBadParse?: boolean;
    }
  ): { node: Note | null; missing: string | null } {
    opts = _.defaults(opts, {
      errorOnEmpty: true,
      isRoot: false,
      errorOnBadParse: true
    });
    // DEBUG: noteProps: {noteProps}
    // TODO: handle errors
    let noteProps: NoteRawProps;
    try {
      noteProps = mdFile2NodeProps(path.join(store.opts.root, ent.fpath));
    } catch (err) {
      logger.error({ ctx: "toNode", ent, opts, err });
      if (opts.errorOnBadParse) {
        throw err;
      } else {
        const errorMsg = {
          status: "BAD_PARSE",
          ent,
          err
        };
        this.errors.push(errorMsg);
        return { node: null, missing: null };
      }
    }
    // OPT
    //logger.debug({ ctx: "toNode:mdFile2NodeProps:post", noteProps })
    // let parent: Note | null;
    let parentPath: string | null = null;
    let missing: null | string = null;
    // if (_.isNull(noteProps.parent)) {
    parentPath = DNodeUtils.dirName(ent.prefix);
    const parent = _.find(parents, p => p.path === parentPath) || null;
    // } else {
    //   parent = _.find(parents, p => p.id === noteProps.parent) || null;
    // }
    // error checking
    if (!parent && !opts.isRoot) {
      const errorMsg = {
        status: "NO_PARENT_PATH",
        ent,
        parentPath
      };
      this.errors.push(errorMsg);
      // should not be the case
      // if (!parentPath) {
      //   throw Error(`no parent path found for ${JSON.stringify(errorMsg)}`);
      // }
      missing = parentPath;
      if (opts.errorOnEmpty) {
        throw new Error(JSON.stringify(errorMsg));
      }
    }
    // children not looked at when parsing since we build iteratively based on parent
    const note = new Note({ ...noteProps, parent, children: [] });
    if (parent) {
      parent.addChild(note);
    }
    // OPT
    //logger.debug({ ctx: "toNode:exit", note: note.toRawProps(), missing })
    return { node: note, missing };
  }

  _parseSchema(fpath: string): SchemaRawProps[] {
    const fname = path.parse(fpath).name;
    const root = this.store.opts.root;

    const schemaOpts: SchemaRawOpts[] = YAML.parse(
      fs.readFileSync(path.join(root, fpath), "utf8")
    );
    const schemaProps = schemaOpts.map(o =>
      SchemaNodeRaw.createProps({ ...o, fname })
    );
    return schemaProps;
  }

  parseSchema(data: string[]): SchemaRawProps[] {
    if (_.isEmpty(data)) {
      return [];
    }
    return data.map(fpath => this._parseSchema(fpath)).flat();
  }

  /**
   * Returns list of notes withou parent/child information
   * @param data
   */
  parse(data: string[]): Note[] {
    if (_.isEmpty(data)) {
      return [];
    }
    // used to reconstruct missing notes
    const nodesStoreByFname: NoteDict = {};
    const store = this.store;
    const fileMetaDict: FileMetaDict = getFileMeta(data);
    // logger.debug({ ctx: "parse:getFileMeta:post", fileMetaDict })
    const root = fileMetaDict[1].find(n => n.fpath === "root.md") as FileMeta;
    if (!root) {
      return [];
    }
    const { node: rootNode } = this.toNode(root, [], store, {
      isRoot: true,
      errorOnBadParse: this.opts.errorOnBadParse
    }) as { node: Note };
    nodesStoreByFname["root"] = rootNode;

    // domain root
    let lvl = 2;
    let prevNodes: Note[] = fileMetaDict[1]
      // don't count root node, handle separately
      .filter(n => n.fpath !== "root.md")
      .map(ent =>
      // first layer, no parent, expected
      {
        const { node } = this.toNode(ent, [], store, {
          isRoot: true
        });
        return node;
      }
      ) as Note[];
    prevNodes.forEach(n => {
      rootNode.addChild(n);
      nodesStoreByFname[n.fname] = n;
    });

    // domain root children
    while (_.has(fileMetaDict, lvl)) {
      // OPT: slow
      // logger.debug({
      //     ctx: "parse:while",
      //     prevNodes: prevNodes.map(n => n.toRawProps()),
      // })
      const currNodes = fileMetaDict[lvl]
        // eslint-disable-next-line no-loop-func
        .map(ent => {
          // ignore root.schema and other such files
          if (globMatch(["root.*"], ent.fpath)) {
            return null;
          }
          const { node, missing } = this.toNode(ent, prevNodes, store, {
            errorOnEmpty: this.opts.errorOnEmpty,
            errorOnBadParse: this.opts.errorOnBadParse
          });
          if (missing) {
            const closetParent = DNodeUtils.findClosestParent(
              (node as Note).logicalPath,
              nodesStoreByFname
            );
            const stubNodes = NoteUtils.createStubNotes(
              closetParent as Note,
              node as Note
            );
            stubNodes.forEach(ent2 => {
              nodesStoreByFname[ent2.fname] = ent2;
            });
            // TODO: add stub nodes to dict and return dict
            this.missing.add(missing);
          }
          return node;
        })
        .filter(Boolean) as Note[];
      lvl += 1;
      currNodes.forEach(n => {
        nodesStoreByFname[n.fname] = n;
      });
      // TODO: remove
      prevNodes = currNodes;
    }
    return _.values(nodesStoreByFname);
  }

  report() {
    const { errors, missing } = this;
    return {
      numErrors: errors.length,
      errors,
      missing: Array.from(missing)
    };
  }
}
