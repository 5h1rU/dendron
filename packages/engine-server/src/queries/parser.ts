import type { Node, Parent } from "unist";

enum DendronQLNodeType {
  FNAME = "fname",
  VAULT_NAME = "vault_name",
  VARIABLE = "variable",
  CONSTANT = "constant",
  FUNCTION = "function",
}

type DendronNode = Node & {
  type: DendronQLNodeType;
};

type DendronQLParser<T extends DendronNode> = (text: string) => T | undefined;

type DendronQLFname = DendronNode & {
  type: DendronQLNodeType.FNAME;
  value: string;
};

/** Parses a note name. */
const parseFname: DendronQLParser<DendronQLFname> = (text) => {
  // Any text without whitespace or square brackets
  const match = text.match(/^([^[\]\s]+)$/);
  if (match) {
    return {
      type: DendronQLNodeType.FNAME,
      value: match[1],
    };
  }
  return;
};

type DendronQLVaultName = DendronNode & {
  type: DendronQLNodeType.VAULT_NAME;
  value: string;
};

/** Parses a vault name. */
const parseVaultName: DendronQLParser<DendronQLVaultName> = (text) => {
  // Any text without whitespace or square brackets
  const match = text.match(/^([^[\]\s]+)$/);
  if (match) {
    return {
      type: DendronQLNodeType.VAULT_NAME,
      value: match[1],
    };
  }
  return;
};

type DendronQLValue = DendronNode &
  (
    | {
        type: DendronQLNodeType.VARIABLE;
        value: string;
      }
    | {
        type: DendronQLNodeType.CONSTANT;
        value: string | number;
      }
  );

/** Parses a value, which may be a constant or a variable. */
const parseValue: DendronQLParser<DendronQLValue> = (text) => {
  // Any string which is text without whitespace or square brackets surrounded by quotes,
  // or any number which is 123 or 123.45 or .45
  const match = text.match(
    /^(["'](?<str>[^[\]\s]+)["']|(?<var>[^[\].\d\s][^[\]\s]*)|(?<num>\d+([.]\d*)?|[.]\d+))$/
  );
  if (match && match.groups) {
    if (match.groups.str) {
      return { type: DendronQLNodeType.CONSTANT, value: match.groups.str };
    }
    if (match.groups.num) {
      return {
        type: DendronQLNodeType.CONSTANT,
        value: Number(match.groups.num),
      };
    }
    if (match.groups.var) {
      return { type: DendronQLNodeType.VARIABLE, value: match.groups.var };
    }
  }
  return;
};

type DendronQLFunction = DendronNode & {
  name: string;
};

const DendronQLBuiltins = {
  OR: "or",
  AND: "and",
  LINK: "link",
  TO: "to",
  FROM: "from",
  CONNECTED: "connected",
  CHILDREN: "children",
  PARENT: "parent",
  DESCENDANTS: "descendants",
  ANCESTORS: "ancestors",
  VAULT: "vault",
};

type DendronQLMultiExpression = DendronNode & {};

type DendronQLExpression = DendronNode & {};
