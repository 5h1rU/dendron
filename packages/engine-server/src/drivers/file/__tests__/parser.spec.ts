import { EngineTestUtils } from "@dendronhq/common-server";
import { FileParserUtils } from "../parser";

describe("schema parser ", () => {
  let root: string;

  beforeEach(async () => {
    root = await EngineTestUtils.setupStoreDir({
      storeDirSrc: "engine-server.parser",
    });
    console.log("done");
  });

  test("basic", () => {
    console.log(root);
    const schemas = FileParserUtils.parseSchemaFile("foo.schema.yml", { root });
    expect(schemas).toMatchSnapshot("schemas");
  });
});
