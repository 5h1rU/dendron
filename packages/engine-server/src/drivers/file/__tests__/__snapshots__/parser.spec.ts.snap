// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema parser  basic: schemas 1`] = `
Array [
  Object {
    "body": "",
    "children": Array [
      "bar",
    ],
    "created": 1598123040495,
    "custom": Object {},
    "data": Object {},
    "desc": "",
    "fname": "foo.schema",
    "id": "foo",
    "parent": "root",
    "stub": false,
    "title": "foo",
    "updated": 1598123040495,
  },
]
`;
